### PR TITLE
feat: Native Broadcast nested loop join support

### LINKED ISSUE #198: Support BroadcastNestedLoopJoinExec
### What is the problem the feature request solves?

Datafusion supports 
Cross and NestedLoop joins as well : https://docs.rs/datafusion-physical-plan/36.0.0/datafusion_physical_plan/joins/index.html

It will really nice if we can add support for it like Hash and SortMergeJoin.

### Describe the potential solution

_No response_

### Additional context

SHJ pr 
- https://github.com/apache/arrow-datafusion-comet/pull/194
SMJ pr 
- https://github.com/apache/arrow-datafusion-comet/pull/178

### PR DESCRIPTION
**Pull request by coderfender**
## Which issue does this PR close?

Closes #198

Wires Spark's `BroadcastNestedLoopJoinExec` to DataFusion's `NestedLoopJoinExec`.

## Rationale for this change

Native support for BroadcastNestedLoopJoin to improve native performance. 

## How are these changes tested?

1. Unit tests in CometJoinSuite 
2. Setup BNLJ benches

### CODE REVIEW

**Comment by coderfender** [comment]
Bunch TPCDS queries seem to be using BNLJ warranting golden file regen

**Comment by coderfender** [comment]
Benchmarks (local M5 pro) (Edited after running benchmarks with JIT disabled) :


```
Running benchmark: range join (BETWEEN)
  Running case: Spark
  Stopped after 23 iterations, 2043 ms
  Running case: Comet
  Stopped after 33 iterations, 2051 ms

OpenJDK 64-Bit Server VM 17.0.16+8-LTS on Mac OS X 26.3.2
Apple M5 Pro
range join (BETWEEN):                     Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Spark                                                73             89          10         14.4          69.7       1.0X
Comet                                                49             62          11         21.4          46.8       1.5X

Running benchmark: inequality join (>)
  Running case: Spark
  Stopped after 21 iterations, 2038 ms
  Running case: Comet
  Stopped after 34 iterations, 2028 ms

OpenJDK 64-Bit Server VM 17.0.16+8-LTS on Mac OS X 26.3.2
Apple M5 Pro
inequality join (>):                      Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Spark                                                78             97          19         13.4          74.4       1.0X
Comet                                                48             60           6         21.8          46.0       1.6X

Running benchmark: left outer non-equi
  Running case: Spark
  Stopped after 27 iterations, 2028 ms
  Running case: Comet
  Stopped after 35 iterations, 2028 ms

OpenJDK 64-Bit Server VM 17.0.16+8-LTS on Mac OS X 26.3.2
Apple M5 Pro
left outer non-equi:                      Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Spark                                                63             75          11         16.6          60.1       1.0X
Comet                                                42             58          16         25.2          39.7       1.5X
```

**Comment by coderfender** [comment]
TODO : Build vs Probe swap to DF's BNLJ op

**Comment by coderfender** [comment]
Investigating test failures

File: spark/src/main/scala/org/apache/spark/sql/comet/operators.scala [hunk 1]
```diff
@@ -1924,6 +1924,158 @@ trait CometHashJoin {
...
+  override def enabledConfig: Option[ConfigEntry[Boolean]] = {
+    Some(CometConf.COMET_EXEC_BROADCAST_NESTED_LOOP_JOIN_ENABLED)
+  }
+
+  private val broadcastBuildReplicationReason =
+    "BNLJ combinations that emit per-build-row results need a cross-partition merge that" +
+      " DataFusion's NestedLoopJoinExec does not provide. Affects: LeftOuter+BuildLeft," +
+      " RightOuter+BuildRight, FullOuter, LeftSemi+BuildLeft, LeftAnti+BuildLeft."
+
+  override def getSupportLevel(op: BroadcastNestedLoopJoinExec): SupportLevel =
```
Comment (mbutrovich) [inline spark/src/main/scala/org/apache/spark/sql/comet/operators.scala:1993]: The support matrix here lines up exactly with Spark's `BroadcastNestedLoopJoinExec.outputPartitioning` supported set, and the reasoning in `broadcastBuildReplicationReason` is spot on. Every case returns either `Compatible(None)` or `Unsupported(...)`, never `Incompatible`. Since `isOperatorEnabled` only consults the allow-incompat config on the `Incompatible` branch, the `getOperatorAllowIncompatConfigKey("BroadcastNestedLoopJoinExec")` line in the `CometJoinSuite` override is dead. Is that intentional, or leftover?

File: spark/src/main/scala/org/apache/spark/sql/comet/operators.scala [hunk 2]
```diff
@@ -1924,6 +1924,158 @@ trait CometHashJoin {
   }
 }
 
+case class CometBroadcastNestedLoopJoinExec(
```
Comment (mbutrovich) [inline spark/src/main/scala/org/apache/spark/sql/comet/operators.scala:1927]: Should this define a `metrics` map like `CometMetricNode.hashJoinMetrics` does for BHJ (and `sortMergeJoinMetrics` for SMJ)? As written it inherits `baselineMetrics`, so the native NLJ's build/probe/join timing metrics won't surface in the Spark UI.

File: spark/src/test/scala/org/apache/comet/exec/CometJoinSuite.scala [hunk 3]
```diff
@@ -38,7 +38,10 @@ class CometJoinSuite extends CometTestBase {
   override protected def test(testName: String, testTags: Tag*)(testFun: => Any)(implicit
       pos: Position): Unit = {
     super.test(testName, testTags: _*) {
-      withSQLConf(CometConf.COMET_EXEC_SHUFFLE_ENABLED.key -> "true") {
+      withSQLConf(
+        CometConf.COMET_EXEC_SHUFFLE_ENABLED.key -> "true",
+        CometConf.COMET_EXEC_BROADCAST_NESTED_LOOP_JOIN_ENABLED.key -> "true",
+        CometConf.getOperatorAllowIncompatConfigKey("BroadcastNestedLoopJoinExec") -> "true") {
```
Comment (mbutrovich) [inline spark/src/test/scala/org/apache/comet/exec/CometJoinSuite.scala]: Similar to the other comment in operators.scala. This allow-incompat config is never read for BNLJ. `getSupportLevel` only ever returns `Compatible(None)` or `Unsupported(...)`, never `Incompatible`, and `isOperatorEnabled` only consults allow-incompat on the `Incompatible` branch. Can this be dropped, or was an `Incompatible` case intended?

File: spark/src/test/scala/org/apache/comet/exec/CometJoinSuite.scala [hunk 4]
```diff
@@ -702,4 +705,102 @@ class CometJoinSuite extends CometTestBase {
       }
     }
   }
+
+  test("BroadcastNestedLoopJoin with unequal filter") {
```
Comment (mbutrovich) [inline spark/src/test/scala/org/apache/comet/exec/CometJoinSuite.scala:706]: The compatible matrix is well covered. Could we add a couple of result-only `checkSparkAnswer` cases for the unsupported combinations (`FullOuter`, `LeftOuter` with `BuildLeft`) to confirm they fall back to Spark and still match? A broadcast-reuse case (same relation feeding two BNLJs under AQE) would also be reassuring given reuse is handled generically rather than in this operator. Minor: no coverage for a join condition referencing nullable or null-producing columns, and no left-outer-without-condition case.

**Review by mbutrovich — CHANGES_REQUESTED** [review]
First round of feedback. Thanks @coderfender! This is looking really good!

File: spark/src/main/scala/org/apache/spark/sql/comet/operators.scala (hunk 1 above)
Comment (coderfender) [reply spark/src/main/scala/org/apache/spark/sql/comet/operators.scala:1993]: Agreed. I was initially gating the join behind the config in the test suite to figure out the support matrix through testing. I have since then removed the now redundant config

File: spark/src/main/scala/org/apache/spark/sql/comet/operators.scala (hunk 2 above)
Comment (coderfender) [reply spark/src/main/scala/org/apache/spark/sql/comet/operators.scala:1927]: Verified DF's `NestedLoopJoinExec` uses the same `BuildProbeJoinMetrics` shape as `HashJoinExec` So instead of duplicating the metric map, renamed `hashJoinMetrics` to `joinMetrics` in `CometMetricNode.scala`, updated the two existing call sites (BHJ + SHJ case classes), and added a new override CometBroadcastNestedLoopJoinExec to leverage the same metric report function

File: spark/src/test/scala/org/apache/comet/exec/CometJoinSuite.scala (hunk 3 above)
Comment (coderfender) [reply spark/src/test/scala/org/apache/comet/exec/CometJoinSuite.scala]: Addressed this. Thank you

File: spark/src/test/scala/org/apache/comet/exec/CometJoinSuite.scala (hunk 4 above)
Comment (coderfender) [reply spark/src/test/scala/org/apache/comet/exec/CometJoinSuite.scala:706]: Thank you.
1.  Added LeftOuter + BuildLeft, FullOuter tests which should fall back to spark.
2.  Also added nulls to input data for all the test cases to verify null join key condition as well
3. Broadcast reuse was a great idea ! I added a test to verify that as well

**Review by mbutrovich — APPROVED** [review]
Thanks for working through the first round, this looks ready to me. I re-checked the support matrix against Spark's `BroadcastNestedLoopJoinExec` and the swap path against DataFusion's `NestedLoopJoinExec`, and both hold up.

A few things I verified:

- The `Compatible` set matches exactly the cases where Spark returns `streamed.outputPartitioning`. Every unsupported combo hits Spark's `defaultJoin` or `getMatchedBroadcastRowsBitSet`, which do a cross-partition merge that DF's NLJ cannot reproduce. The `broadcastBuildReplicationReason` is accurate.
- DF's `swap_inputs` correctly swaps the join type, the filter, and reorders output, so the `BuildRight` path is sound and consistent with the existing `HashJoin` branch.
- DF's `NestedLoopJoinExec` emits the same `BuildProbeJoinMetrics` shape as `HashJoinExec`, so the `joinMetrics` rename is correct and metrics will surface.
- `outputPartitioning` is unconditional rather than guarded like Spark's, but that is safe since the operator is only constructed for the `Compatible` set, which is exactly the set where Spark also returns streamed partitioning.

Two minor, non-blocking points:

- `Cross` is a separate `InnerLike` join type, and Spark's BNLJ treats it like `Inner`. As written, an explicit `CROSS JOIN` produces join type `Cross`, hits the `case _` fallback, and falls back to Spark. Comma-joins resolve to `Inner` so they run natively, but the explicit keyword would not. Worth adding `Cross` alongside `Inner` here and in `parse_join_parameters`, or leaving as a follow-up.
- `LeftSingle + BuildRight` is in Spark's streamed set but has a max-one-match guarantee that DF's NLJ does not provide, so the fallback is correct. A one-line comment noting that is intentional would help.

**Comment by coderfender** [comment]
Thank you very much for the approval @mbutrovich

**Comment by coderfender** [comment]
Tagging @parthchandra @comphead @andygrove for additional review

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @coderfender

**Comment by coderfender** [comment]
Thank you for the approval @andygrove
